### PR TITLE
Add extensions functions to Gipsy to support conversion to/from JSON.

### PR DESCRIPTION
**Pull request by cmickeyb**
Added json-to-expression, expression-to-json and validate-json functions
to Gipsy. While this is in preparation for unifying the interpreter API
it also provides useful functions. Note that not all s-expressions can
be converted to JSON. Specifically:

- basic types, string, numbers, booleans are converted as expected
- symbols are not converted at all
- vectors are converted to vectors
- lists must be associative arrays, i.e. a list of two element lists; these are converted to a JSON object

Tests of the new functions are provided in the Gipsy tests directory.

Signed-off-by: Mic Bowman <mic.bowman@intel.com>

### CODE REVIEW

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp [hunk 1]
```diff
@@ -1046,6 +1050,312 @@ static pointer key_value_close(scheme *sc, pointer args)
...
+        if (sc->vptr->is_list(sc, expr))
+        {
+            j_value = json_value_init_object();
+            j_object = json_value_get_object(j_value);
+
+            for (pointer next = expr; next != sc->NIL; next = sc->vptr->pair_cdr(next))
+            {
+                pointer elem = sc->vptr->pair_car(next);
+                int length = sc->vptr->list_length(sc, elem);
+                pe::ThrowIf<pe::RuntimeError>(length != 2, "invalid JSON object representation");
```
Comment (g2flyer) [inline common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: Error message is a bit missleading? It's rather an invalid s-expression not representable as JSON?

Comment (g2flyer) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: BTW: to understand that code it is crucial to have seen the comment you had in your PR. That mapping outline should really be in the code itself as mentioned also elsewhere.

**Comment by cmickeyb** [comment]
> Looks in general good and e2e test also succeed (although for very simple tests :-). Main request for changes are adding docu of mapping strategy from PR description into code itself and a comment on empty catch. "outsourcing" in a separate file would be nice but not a show-stopper.

Separate file would, in general be preferred--actually that entire file should be split into components--however, at this point the primary objective is to move towards unifying the interface to the interpreters. I would rather keep the restrictive use of the current implementation so we can focus more on the WASM code rather than spend a lot of time trying to generalize something that is unlikely to be used in its generalized form.

File: common/interpreter/gipsy_scheme/tests/json.scm [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+
+(display "START JSON\n")
+
+(catch error-print
+       (define (simple-expression v) (equal? v (json-to-expression (expression-to-json v))))
+
+       (test "single value tests")
+       (assert (simple-expression 1) "failed to convert integer")
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/tests/json.scm]: given that this code is specifically adding scheme functions, the c unit tests seems like a substantial overkill. the json.scm file in the tests directory can be extended as necessary.

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp [hunk 3]
```diff
@@ -1046,6 +1050,312 @@ static pointer key_value_close(scheme *sc, pointer args)
...
+    case JSONNull:
+        return sc->NIL;
+
+    case JSONError:
+    default:
+        pe::ThrowIf<pe::RuntimeError>(true, "unknown error in JSON processing");
+    }
+
+    // should never reach here
+    return sc->NIL;
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: or just remove the code completely. there is, in theory, no way to reach that location in the code and the compiler knows it & doesn't complain.

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp [hunk 4]
```diff
@@ -1046,6 +1050,312 @@ static pointer key_value_close(scheme *sc, pointer args)
...
+    try {
+        JsonValue parsed(json_parse_string(message.c_str()));
+        pe::ThrowIfNull(parsed.value, "failed to parse JSON expression");
+
+        return json_to_expression_r(sc, parsed);
+    }
+    catch (pdo::error::Error& e) {
+        return scheme_return_error_s(sc, format_error_message(e));
+    }
+    catch (...) {
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: yes. it should be completely safe. the error that is returned in the next line is the capture all error (and there is no way to get an actual exception message from the "..." exceptions... there are only a couple that are not derived from the error exception that are caught there)

[the remark this answers is not in the document]
Comment (bvavala) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: > yes. it should be completely safe. the error that is returned in the next line is the capture all error (and there is no way to get an actual exception message from the "..." exceptions... there are only a couple that are not derived from the error exception that are caught there)

I am not sure that `pdo::error` block also handles the `bad_alloc` exception. I think it does **not**.
In this case, we may end up translating the `bad_alloc` in a scheme error. Then the error function might set the `no_memory` scheme flag. So we should abort and shutdown the interpreter.
How about throwing a runtime error exception as in https://github.com/hyperledger-labs/private-data-objects/blob/f439d18f6d118f1b2a51c870a9c39f9114fb326a/common/interpreter/gipsy_scheme/GipsyInterpreter.cpp#L150 ?

[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: all of these get caught correctly in the expected way in the same style that we catch them for all other scheme extensions.

File: common/interpreter/gipsy_scheme/tests/README.md [hunk 5]
```diff
@@ -8,35 +8,11 @@ https://creativecommons.org/licenses/by/4.0/
...
-
-`make TINYSCHEME=~/dev/packages/tinyscheme-1.41/ pcontract.so`
-
 ## Run the Tests ##
 
 From the tests directory, the tests can be run with the following
 command:
 
-`LD_LIBRARY_PATH=.. tinyscheme -1 test.scm`
+`${PDO_SOURCE_ROOT}/contracts/bin/gipsyscheme.sh test.scm`
```
Comment (bvavala) [inline common/interpreter/gipsy_scheme/tests/README.md]: This should be `${PDO_SOURCE_ROOT}/contracts/bin/gipsyscheme test.scm`

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp [hunk 6]
```diff
@@ -1046,6 +1050,312 @@ static pointer key_value_close(scheme *sc, pointer args)
 }
 #endif
 
+/* ----------------------------------------------------------------- */
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: comments added to code.

File: common/interpreter/gipsy_scheme/tests/json.scm (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/tests/json.scm]: additional tests added to json.scm

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: removed unreachable code.

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: this is not broken & consistent with all other scheme extensions functions.

File: common/interpreter/gipsy_scheme/SchemeExtensions.cpp (hunk 1 above)
Comment (cmickeyb) [reply common/interpreter/gipsy_scheme/SchemeExtensions.cpp]: Updated

**Pull request merged**